### PR TITLE
controller: Add non-concurrent commissioning

### PR DESCRIPTION
**Pull request by wqx6**
### Problem
According to SPEC, for non-concurrent commissioning we should terminate the commissioning channel successful step 12 (trigger joining of operational network at Commissionee) and then find the operational node, establish CASE Session, and send commissioning-complete.

### Solution
Release the commissionee device proxy after receiving the success response of connect-network.

### Test
Tested chip-tool, used it to pairing ESP32 with BLE.
When using concurrent commissioning, the chip-tool will close the BLE connection after sending commissioning-complete.
When using non-concurrent commissioning, the chip-tool will close the BLE connection after receiving the success response of connect-network.

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -1689,19 +1717,31 @@ void DeviceCommissioner::OnDeviceConnectedFn(void * context, Messaging::Exchange
         return;
     }
 
-    if (commissioner->mDeviceBeingCommissioned == nullptr ||
-        commissioner->mDeviceBeingCommissioned->GetDeviceId() != sessionHandle->GetPeer().GetNodeId())
+    if (commissioner->mCommissioningDelegate == nullptr)
     {
-        // Not the device we are trying to commission.
         return;
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1722]: This seems like a logic change: before we would proceed even if `mCommissioningDelegate` was null.  Why the change?

File: src/controller/CHIPDeviceController.cpp [hunk 2]
```diff
@@ -2442,7 +2482,7 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
     break;
     case CommissioningStage::kFindOperational: {
         // If there is an error, CommissioningStageComplete will be called from OnDeviceConnectionFailureFn.
-        auto scopedPeerId = GetPeerScopedId(proxy->GetDeviceId());
+        auto scopedPeerId = GetPeerScopedId(params.GetRemoteNodeId().Value());
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:2485]: What guarantees that `GetRemoteNodeId()` has a value here?  Should probably be documented.

File: src/controller/CommissioningDelegate.h [hunk 3]
```diff
@@ -457,6 +464,7 @@ class CommissioningParameters
     Optional<bool> mAttemptWiFiNetworkScan;
     Optional<bool> mAttemptThreadNetworkScan; // This automatically gets set to false when a ThreadOperationalDataset is set
     Optional<bool> mSkipCommissioningComplete;
+    Optional<bool> mNonConcurrentCommissioning;
```
Comment (bzbarsky-apple) [inline src/controller/CommissioningDelegate.h:467]: So here is a question: does this really need to be an Optional?  Or can it just be a boolean?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Comment by s07641069** [comment]
@wqx6 Telink did initial testing of non-concurrent commissioning implementation and it works fine on out side with chip-tool when we use "--non-concurrent" flag.
But it doesn't work with Google and Apple Homes at all.
Do you have some ideas if non-concurrent commissioning will be supported by Google and Apple Homes?

**Comment by bzbarsky-apple** [comment]
Actually, why is any of this requiring client-side changes at all, now that I think about it?  The _server_ should be doing its non-concurrent bits, but why does the _client_ want to tear down the PASE session early.  That does not look right to me.  @wqx6

File: src/controller/AutoCommissioner.cpp [hunk 4]
```diff
@@ -661,7 +677,8 @@ DeviceProxy * AutoCommissioner::GetDeviceProxyForStep(CommissioningStage nextSta
 CHIP_ERROR AutoCommissioner::PerformStep(CommissioningStage nextStage)
 {
     DeviceProxy * proxy = GetDeviceProxyForStep(nextStage);
-    if (proxy == nullptr)
+    if ((!mParams.GetNonConcurrentCommissioning().ValueOr(false) || nextStage != CommissioningStage::kFindOperational) &&
```
[the remark this answers is not in the document]
Comment (wqx6) [reply src/controller/AutoCommissioner.cpp]: Because when using non-concurrent commissioning, the device_proxy will be null in step FindOperational, the PASE Session device proxy should be released after step `Joining Operational Network`.

File: src/controller/AutoCommissioner.cpp [hunk 5]
```diff
@@ -432,12 +438,15 @@ Optional<System::Clock::Timeout> AutoCommissioner::GetCommandTimeout(DeviceProxy
...
     }
 
-    // Adjust the timeout for our session transport latency, if we have access
-    // to a session.
-    auto sessionHandle = device->GetSecureSession();
-    if (sessionHandle.HasValue())
+    if (!mParams.GetNonConcurrentCommissioning().ValueOr(false))
     {
-        timeout = sessionHandle.Value()->ComputeRoundTripTimeout(timeout);
+        // Adjust the timeout for our session transport latency, if we have access
```
[the remark this answers is not in the document]
Comment (wqx6) [reply src/controller/AutoCommissioner.cpp:440]: Change the condition to that device is not null.

File: src/controller/CHIPDeviceController.cpp [hunk 6]
```diff
@@ -1639,18 +1639,46 @@ void DeviceCommissioner::CommissioningStageComplete(CHIP_ERROR err, Commissionin
...
-        // operations) when we are not in fact commissioning anything.  Just
-        // ignore it.
-        return;
+        if (mDeviceBeingCommissioned != nullptr)
+        {
+            nodeId = mDeviceBeingCommissioned->GetDeviceId();
+            proxy  = mDeviceBeingCommissioned;
+            if (mCommissioningStage == kThreadNetworkEnable || mCommissioningStage == kWiFiNetworkEnable)
+            {
+                ReleaseCommissioneeDevice(reinterpret_cast<CommissioneeDeviceProxy *>(proxy));
```
[the remark this answers is not in the document]
Comment (wqx6) [reply src/controller/CHIPDeviceController.cpp:1652]: Both the `mDeviceBeingCommissioned` and `proxy` is type `DeviceProxy`. we should do the cast since the release function needs a `CommissioneeDeviceProxy` input.

File: src/controller/CHIPDeviceController.cpp [hunk 7]
```diff
@@ -1639,18 +1639,46 @@ void DeviceCommissioner::CommissioningStageComplete(CHIP_ERROR err, Commissionin
...
+        }
+        else if (mCommissioningStage == kFindOperational &&
+                 mCommissioningDelegate->GetCommissioningParameters().GetRemoteNodeId().HasValue())
+        {
+            nodeId = mCommissioningDelegate->GetCommissioningParameters().GetRemoteNodeId().Value();
+            proxy  = nullptr;
+        }
+        else
+        {
+            return;
```
[the remark this answers is not in the document]
Comment (wqx6) [reply src/controller/CHIPDeviceController.cpp:1665]: Added

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (wqx6) [reply src/controller/CHIPDeviceController.cpp:1722]: Before the changes, we will do nothing if `mCommissioningDelegate` is null. so I moved the check of `mCommissioningDelegate` to here.  If `mCommissioningDelegate` is null, the function will directly return after checking the nodeId.

File: src/controller/CHIPDeviceController.cpp (hunk 2 above)
Comment (wqx6) [reply src/controller/CHIPDeviceController.cpp]: Add the check for the node ID.

File: src/controller/CommissioningDelegate.h (hunk 3 above)
Comment (wqx6) [reply src/controller/CommissioningDelegate.h]: Changed it to boolean

File: examples/chip-tool/commands/pairing/PairingCommand.h [hunk 8]
```diff
@@ -140,6 +140,7 @@ class PairingCommand : public CHIPCommand,
         }
 
         AddArgument("timeout", 0, UINT16_MAX, &mTimeout);
+        AddArgument("non-concurrent", 0, 1, &mNonConcurrent);
```
[the remark this answers is not in the document]
Comment (wqx6) [reply examples/chip-tool/commands/pairing/PairingCommand.h]: Added

**Comment by wqx6** [comment]
> why does the client want to tear down the PASE session early
Because the SPEC says
```
In concurrent connection commissioning flow the commissioning channel SHALL terminate after successful step
15 (CommissioningComplete command invocation). In non-concurrent connection commissioning flow the
commissioning channel SHALL terminate after successful step 12 (trigger joining of operational network at
Commissionee).
```
So I tear down the PASE session when finishing the step `join the operational network`.

**Comment by msandstedt** [comment]
I would echo @bzbarsky-apple's comment.  The server knows whether it can support concurrent commissioning or not.  So why doesn't it just close the session on its end if necessary?

File: src/controller/AutoCommissioner.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/controller/AutoCommissioner.cpp]: > the PASE Session device proxy should be released after step Joining Operational Network

Why?

File: src/controller/CHIPDeviceController.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/controller/CHIPDeviceController.cpp:1652]: Why can't we make `mDeviceBeingCommissioned` a `CommissioneeDeviceProxy *`?  And if we can't, how do we know it's safe to cast it to that?

**Comment by bzbarsky-apple** [comment]
Exactly.  For non-concurrent commissioning the _server_ should just close the BLE connection as needed, I would think.  The client closing things here requires the client to know whether the server supports concurrent commissioning, which it _could_ read via SupportsConcurrentConnection, but this PR is not doing that.  Passing in the "do concurrent commissioning" state out-of-band seems wrong in any case; we should be getting it from the server if we think the client needs to do something special here.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Not clear why any of the client changes are needed, and the chip-tool changes should _definitely_ not be needed.

**Comment by wqx6** [comment]
> Not clear why any of the client changes are needed

The client can know whether the server supports concurrent commissioning by reading the `SupportsConcurrentConnection` attribute. But if this attribute is true, the client should support both concurrent and non-concurrent commissioning. The client should be able to choose the commissioning flow type.

File: src/controller/CHIPDeviceController.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (wqx6) [reply src/controller/CHIPDeviceController.cpp]: Changed to release the device_poxy from the commissionee device pool.

**Comment by bzbarsky-apple** [comment]
>  But if this attribute is true, the server should support both concurrent and non-concurrent commissioning

Why?

> If the client cannot choose the commissioning flow, how can we do non-concurrent commissioning for the devices that support both the two commissioning flow.

Why would we want to do that?  The only reason non-concurrent commissioning exists is to handle servers that can't speak both BLE and another radio technology at the same time.  It's strictly worse than concurrent commissioning in all sorts of ways; the only reason you would use it is if you're forced to do it.

**Comment by wqx6** [comment]
> The only reason non-concurrent commissioning exists is to handle servers that can't speak both BLE and another radio technology at the same time.

Understand, close this PR.

**Comment by wqx6** [comment]
@bzbarsky-apple I still have a question. If a server doesn't support concurrent commissioning because it can't speak both BLE and another radio technology at the same time. When should it close the BLE connection? If it closes the BLE when it receives the command `ConnectNetwork`, how can it sends `ConnectNetworkResponse` since the PASE Session has been terminated? If it closes the BLE after it sends `ConnectNetworkResponse`, there must be a period when it speaks both BLE and another radio technology at the same time.

**Comment by bzbarsky-apple** [comment]
> When should it close the BLE connection? If it closes the BLE when it receives the command ConnectNetwork, how can it sends ConnectNetworkResponse since the PASE Session has been terminated?

The relevan spec bits seem to be these, in "11.8.8.10. ConnectNetwork Command":

> Success or failure of this command SHALL be communicated by the ConnectNetworkResponse command

> When non-concurrent commissioning is being used by a Commissioner or Administrator, it is possible that the only method to determine success of the operation is operational discovery of the Node on the new operational network. 

These are obviously mutually contradictory.  Please raise a spec issue to get this sorted?

In practical terms, the only options a non-concurrent commissionnee has are to send ConnectNetworkResponse before trying to connect, or to not send it at all.  And the spec really needs to pin down which of these is the case, so commissioners know what to expect.
